### PR TITLE
Create multiple analysis files for multi-video projects

### PR DESCRIPTION
**Pull request by roomrys**
### Description
Add a menu (+CLI) option to export multiple HDF5 analysis files for multi-video SLP files.

### Types of changes

- [ ] Bugfix
- [x] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
Exporting analysis HDF5 files for multi-video projects[#729](https://github.com/talmolab/sleap/issues/729)

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/wiki/Developer-Guide) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

File: tests/io/test_convert.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    output_prefix = slp_path.with_suffix("")
+    args = f"--format analysis {slp_path}".split()
+    sleap_convert(args)
+    assert_analysis_existance(output_prefix)
+
+    # Specify output and retest
+    output_prefix = tmpdir.with_name("prefix")
+    args = f"--format analysis -o {output_prefix} {slp_path}".split()
+    sleap_convert(args)
+    assert_analysis_existance(output_prefix)
```
Comment (roomrys) [inline tests/io/test_convert.py:41]: lets make this into a `call_and_assert_with_output()` function

**Review by roomrys — COMMENTED** [review]
One "problem", a few improvements.

File: sleap/gui/commands.py [hunk 2]
```diff
@@ -956,25 +953,53 @@ class ExportAnalysisFile(AppCommand):
...
 
-        filename, selected_filter = FileDialog.save(
-            context.app,
-            caption="Export Analysis File...",
-            dir=default_name,
-            filter="SLEAP Analysis HDF5 (*.h5)",
-        )
+        output_paths = []
+        analysis_videos = []
+        for video in videos:
```
Comment (talmo) [inline sleap/gui/commands.py:984]: Not a fan of doing this in a loop here. It will pop up a dialog for every single video in the project -- some of which could have 30+.

If there is more than one video, this command should instead show a `FileDialog.openDir` to prompt the user to pick a directory to save everything to. They lose the ability to customize the filename, but that's way better than clicking "Save" 30+ times.

**Review by talmo — CHANGES_REQUESTED** [review]
One more change needed for UX, but otherwise LGTM

File: sleap/gui/commands.py (hunk 2 above)
Comment (roomrys) [reply sleap/gui/commands.py:1011]: Agreed - this could be very annoying. I changed this to open a `FileDialog.openDir` when multiple videos exist, otherwise use `FileDialog.save` for single video projects.

File: sleap/gui/commands.py [hunk 3]
```diff
@@ -956,25 +953,53 @@ class ExportAnalysisFile(AppCommand):
...
+            all_videos = context.labels.videos
+        else:
+            all_videos = [context.state["video"] or context.labels.videos[0]]
+
+        # Check for labeled frames in each video
+        videos = []
+        for video in all_videos:
+            lfs = labels.get(video)
+            if len(lfs) != 0:
+                videos.append(video)
```
Comment (roomrys) [inline sleap/gui/commands.py:986]: simplify with `videos = [video for video in all_videos if len(labels.get(video)) != 0]`

File: sleap/gui/commands.py [hunk 4]
```diff
@@ -956,25 +954,76 @@ class ExportAnalysisFile(AppCommand):
...
+            # Allow user to specify the filename
+            use_default = False
+            dirname = str(fn.parent)
+        else:
+            # Allow user to specify directory, but use default filenames
+            use_default = True
+            dirname = FileDialog.openDir(
+                context.app,
+                caption="Select Folder to Export Analysis Files...",
+                dir=str(PurePath(default_name).parent),
```
Comment (roomrys) [inline sleap/gui/commands.py:1003]: Simplify with `dir=str(fn.parent)`

**Review by roomrys — COMMENTED** [review]
A few small suggestions for improvement.

File: tests/gui/test_commands.py [hunk 5]
```diff
@@ -63,3 +69,154 @@ def test_get_new_version_filename():
...
+    shutil.copy(small_robot_mp4_vid.backend.filename, tmpdir / "session2" / "video.mp4")
+
+    labels.videos[0].backend.filename = str(tmpdir / "session1" / "video.mp4")
+    labels.videos[1].backend.filename = str(tmpdir / "session2" / "video.mp4")
+
+    params = {"all_videos": True}
+    okay = ExportAnalysisFile_ask(context=context, params=params)
+    assert okay == True
+    ExportAnalysisFile.do_action(context=context, params=params)
+    assert_videos_written(num_videos=2)
```
Comment (talmo) [inline tests/gui/test_commands.py:213]: One more edge case to handle: paths are different, but video filename is the same. This is a common scenario where we have different folder names associated with each video (e.g., session timestamp) but the filename itself is the same.

In this case, the current implementation will overwrite the analysis files. There's a design decision to be made here, but I think the easiest thing would be to just name the output files by their video index within the project.

 See also notes on the metadata that should be added in `write_tracking_h5.py`

File: sleap/info/write_tracking_h5.py [hunk 6]
```diff
@@ -8,10 +7,15 @@
...
-* "node_names"         shape: nodes
+* "track_occupancy"    (shape: tracks * frames)
+* "tracks"             (shape: frames * nodes * 2 * tracks)
+* "track_names"        (shape: tracks)
+* "node_names"         (shape: nodes)
+* "edge_names"         (shape: nodes - 1)
+* "edge_inds"          (shape: nodes - 1)
+* "point_scores"       (shape: frames * nodes * tracks)
+* "instance_scores"    (shape: frames * tracks)
+* "tracking_scores"    (shape: frames * tracks)
```
Comment (talmo) [inline sleap/info/write_tracking_h5.py:18]: Also include:

- `"labels_path"`: Path to the source .slp file (if available from GUI context)
- `"video_path"`: Path to source video (`video.backend.filename`)
- `"video_ind"`: Scalar integer index of the video within the source labels

Also nice to have:
- `"provenance"`: Copied from `labels.provenance`. Might be easier to convert to json string first.

**Review by talmo — CHANGES_REQUESTED** [review]
Important edge case to be considered -- also exposes the need for more metadata.

File: tests/gui/test_commands.py (hunk 5 above)
Comment (roomrys) [reply tests/gui/test_commands.py:213]: I believe the naming should be some-what intuitive/memorable. How about we compromise with a name of `<project_prefix>.<video_idx>_<video_name>.analysis.h5`?

**Review by talmo — APPROVED** [review]
LGTM!
